dtm0/service.c

```diff
@@ -0,0 +1,648 @@
+	       remote_ep);
+
+	if (!async)
+		rc = m0_rpc_link_connect_sync(&process->dop_rlink,
+					      M0_TIME_NEVER);
+	else
+		m0_rpc_link_connect_async(&process->dop_rlink,
+					  M0_TIME_NEVER,
+					  &process->dop_service_connect_clink);
+
+	return M0_RC(rc);
+}
+
+M0_INTERNAL int
+m0_dtm0_service_process_disconnect(struct m0_reqh_service *s,
+				   struct m0_fid          *remote_srv)
+{
+	int rc;
+	struct dtm0_process *process = NULL;
+
+	M0_LOG(M0_DEBUG, "dtm0=%p, remote_srv="FID_F, s, FID_P(remote_srv));
+
+	process = dtm0_service_process__lookup(s, remote_srv);
+	if (process == NULL)
+		return M0_RC(-ENOENT);
+
+	rc = m0_rpc_link_disconnect_sync(
+		&process->dop_rlink,
+		m0_time_from_now(DTM0_DISCONNECT_TIMEOUT_SECS, 0));
+
+	if (rc == -ETIMEDOUT) {
+		M0_LOG(M0_WARN, "Disconnect timeout");
+		rc = 0;
+	}
+
+	m0_rpc_link_fini(&process->dop_rlink);
+
+	return M0_RC(rc);
+}
+
+M0_INTERNAL struct m0_rpc_session *
+m0_dtm0_service_process_session_get(struct m0_reqh_service *s,
+				    const struct m0_fid    *remote_srv)
+{
+	struct dtm0_process *process =
+		dtm0_service_process__lookup(s, remote_srv);
+
+	return process == NULL ? NULL : &process->dop_rlink.rlk_sess;
+}
+
+static int dtm0_service__alloc(struct m0_reqh_service           **service,
+			       const struct m0_reqh_service_type *stype,
+			       const struct m0_reqh_service_ops  *ops)
+{
+	struct m0_dtm0_service *s;
+	int                     rc;
+
+	M0_PRE(stype != NULL && service != NULL && ops != NULL);
+
+	M0_ALLOC_PTR(s);
+	if (s == NULL)
+		return M0_ERR(-ENOMEM);
+
+	s->dos_generic.rs_type = stype;
+	s->dos_generic.rs_ops  = ops;
+	rc = dtm0_service__init(s);
+	if (rc == 0)
+		*service = &s->dos_generic;
+
+	return M0_RC(rc);
+}
+
+static int dtm0_service_allocate(struct m0_reqh_service           **service,
+				 const struct m0_reqh_service_type *stype)
+{
+	return dtm0_service__alloc(service, stype, &dtm0_service_ops);
+}
+
+static int dtm_service__origin_fill(struct m0_reqh_service *service)
+{
+	struct m0_conf_service *service_obj;
+	struct m0_conf_obj     *obj;
+	struct m0_confc        *confc = m0_reqh2confc(service->rs_reqh);
+	const char            **param;
+	struct m0_dtm0_service *dtm0 = to_dtm(service);
+
+	/* W/A for UTs */
+	if (!m0_confc_is_inited(confc)) {
+		dtm0->dos_origin = DTM0_ON_VOLATILE;
+		goto out;
+	}
+
+	obj = m0_conf_cache_lookup(&confc->cc_cache, &service->rs_service_fid);
+	if (obj == NULL)
+		return M0_RC(-ENOENT);
+
+	service_obj = M0_CONF_CAST(obj, m0_conf_service);
+
+	if (service_obj->cs_params == NULL) {
+		dtm0->dos_origin = DTM0_ON_VOLATILE;
+		M0_LOG(M0_WARN, "dtm0 is treated as volatile, no parameters given");
+		goto out;
+	}
+
+	for (param = service_obj->cs_params; *param != NULL; ++param) {
+		if (m0_streq(*param, "origin:in-volatile"))
+			dtm0->dos_origin = DTM0_ON_VOLATILE;
+		else if (m0_streq(*param, "origin:in-persistent"))
+			dtm0->dos_origin = DTM0_ON_PERSISTENT;
```

Comment:
Check that only these two allowed params come from @cs_params. If something else then -EINVAL is to be returned.

Reply:
Some other parameters may be passed, here just persistent and volatile checked to set log type. So not only these two params allowed.